File: scripts/export-schemas.ts
/**
 * Export TypeBox schemas as JSON Schema files for taplo (TOML LSP) autocomplete.
 *
 * Usage: bun run scripts/export-schemas.ts
 */
import { mkdirSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { HeddleConfigSchema } from "../src/config/types.ts";
import { HooksConfigSchema } from "../src/hooks/types.ts";

const SCHEMA_DIR = resolve(import.meta.dirname ?? ".", "../schemas");

const schemas = [
	{ name: "config.schema.json", schema: HeddleConfigSchema },
	{ name: "hooks.schema.json", schema: HooksConfigSchema },
] as const;

mkdirSync(SCHEMA_DIR, { recursive: true });

for (const { name, schema } of schemas) {
	const jsonSchema = {
		$schema: "http://json-schema.org/draft-07/schema#",
		...schema,
	};
	const outPath = resolve(SCHEMA_DIR, name);
	writeFileSync(outPath, `${JSON.stringify(jsonSchema, null, "\t")}\n`);
	console.log(`Exported ${name} → ${outPath}`);
}

File: src/bin/export_schemas.rs
//! Export JSON Schemas for Taplo (TOML LSP) autocomplete + validation.
//!
//! Run: `cargo run --bin export-schemas`
//!
//! Mirrors `ts-src/scripts/export-schemas.ts`. Writes `schemas/config.schema.json`
//! and `schemas/hooks.schema.json` from the `JsonSchema`-deriving structs in
//! `src/config/types.rs` and `src/hooks/types.rs`.

use std::fs;
use std::path::PathBuf;

use anyhow::{Context, Result};
use heddle::schema_export::{config_schema, hooks_schema, pretty_schema};

fn schema_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("schemas")
}

fn write(name: &str, body: serde_json::Value) -> Result<()> {
    let path = schema_dir().join(name);
    fs::create_dir_all(path.parent().unwrap())?;
    fs::write(&path, pretty_schema(&body))
        .with_context(|| format!("writing {}", path.display()))?;
    println!("Exported {} -> {}", name, path.display());
    Ok(())
}

fn main() -> Result<()> {
    write("config.schema.json", config_schema())?;
    write("hooks.schema.json", hooks_schema())?;
    Ok(())
}
